tools/build-tool/data/android/integrated-services/clientserver/app/src/main/java/net/sourceforge/castleengine/ServiceClientServer.java

```diff
@@ -0,0 +1,270 @@
+        {
+            if (!messageMap.containsKey(parts[1]))
+                CreateSocket(parts[1], true, null, Integer.parseInt(parts[4]));
+
+            return true;
+        }
+        else if (parts[2].equals("client")) //client protocol host port
+        {
+            if (!messageMap.containsKey(parts[1]))
+                CreateSocket(parts[1], false, parts[4], Integer.parseInt(parts[5]));
+
+            return true;
+        }
+        else if (parts[2].equals("close")) //close clientid
+        {
+            if (parts[3].equals("all"))
+            {
+                ConcurrentHashMap<String, Boolean> tempMap = activeMap.get(parts[1]).activeMap;
+                synchronized (tempMap)
+                {
+                    for (String mapKey : tempMap.keySet())
+                    {
+                        tempMap.replace(mapKey, false);
+                    }
+                }
+                activeMap.get(parts[1]).active = false;
+            }
+            else
+            {
+                activeMap.get(parts[1]).activeMap.replace(parts[3], false);
+                messageMap.get(parts[1]).remove(parts[3]);
+            }
+
+            return true;
+        }
+        else
+            return false;
+    }
+
+    private void CreateSocket(final String key, final Boolean isServer, final String host, final int port)
+    {
+        final ConcurrentHashMap<String, List<String>> socketMessageMap = new ConcurrentHashMap<String, List<String>>();
+        final ConcurrentHashMap<String, Boolean> socketActiveMap = new ConcurrentHashMap<String, Boolean>();
+
+        messageMap.put(key, socketMessageMap);
+        activeMap.put(key, new ServerTuple(socketActiveMap));
+
+        (new Thread(new Runnable()
+            {
+                @Override
+                public void run()
+                {
+                    try
+                    {
+                        if (isServer)
+                        {
+                            Integer listenerId = 0;
+
+                            ServerSocket serverSocket = new ServerSocket(port);
+                            while (activeMap.get(key).active)
+                            {
+                                CreateListenerThread(serverSocket.accept(), key + Integer.toString(listenerId), socketMessageMap, socketActiveMap);
+                                listenerId++;
+                            }   
+
+                            serverSocket.close();
+                            messageMap.remove(key);
+                            activeMap.remove(key);                       
+                        }
+                        else
+                            CreateListenerThread(new Socket(host, port), key + "client", socketMessageMap, socketActiveMap);
+                    }
+                    catch (IOException e)
+                    {
+                        Log.d("ServiceClientServer", e.toString());
+                    }
```

Comment:
Do not use `Log.xxx` in our services. Instead, use only `logError`, `logWarning` or `logInfo` (they are defined in ancestor `ServiceAbstract`), with `CATEGORY` as the 1st parameter, like

```
logError(CATEGORY, e.toString());
```

Underneath, our `logError` calls Android's `Log.e` method. But it also:
- Makes sure that the messages can be nicely filtered when using `castle-engine run` or `adb logcat -s appName:V`. So it automatically prefixes all log messages with application name.
- And it secures from Android log tag being limited to 23 chars.